vllm/sequence.py

```diff
@@ -499,18 +499,29 @@ def get_output_text_to_return(self, buffer_length: int,
             return self.output_text[last_offset:length]
         return ""
 
-    def get_output_token_ids_to_return(self,
-                                       delta: bool) -> GenericSequence[int]:
+    def get_output_token_ids_to_return(
+            self, delta: bool) -> Union[GenericSequence[int], int]:
         """If delta is True, only new tokens since the last call to
         this method are returned"""
         if not delta:
             return self.get_output_token_ids()
-        length = self.get_output_len()
-        last_offset = self._last_token_ids_offset
-        if last_offset < length:
-            self._last_token_ids_offset = length
-            return self.data._output_token_ids[last_offset:]
-        return ()
+
+        prompt_len = self.get_prompt_len()
+        output_len = self.get_output_len()
+
+        # Get the number of new tokens
+        output_last_offset = self._last_output_token_ids_offset
+        num_new_tokens = output_len - self._last_output_token_ids_offset
+        self._last_output_token_ids_offset = output_len
+
+        # Return new tokens
+        if num_new_tokens == 1:
+            # Optimization for single decode token case
+            # (which is what we have most of the time)
+            return self.data._cached_all_token_ids[-1]
+        else:
+            return self.data._cached_all_token_ids[prompt_len +
+                                                   output_last_offset:]
```

Comment:
nit: minor simplification (prompt_len not needed):
```suggestion
        output_len = self.get_output_len()

        # Get the number of new tokens
        output_last_offset = self._last_output_token_ids_offset
        num_new_tokens = output_len - self._last_output_token_ids_offset
        self._last_output_token_ids_offset = output_len

        # Return new tokens
        if num_new_tokens == 1:
            # Optimization for single decode token case
            # (which is what we have most of the time)
            return self.data._cached_all_token_ids[-1]

        return self.data._cached_all_token_ids[-num_new_tokens:]
```

Reply:
Oh nice, changed the code to use -num_new_tokens.